File: app/services/compare.py
"""关系对比 —— 两人结果摘要的维度反差 / 契合度。

纯模板计算,无外部依赖。对比码 = 对方结果 id,经 /api/results/{id}/public 公开。
"""

from app.models.result import Result


def public_summary(r: Result) -> dict:
    """抽取可公开分享的摘要(不含 user_id / session_id)。"""
    profile = r.profile or {}
    return {
        "id": r.id,
        "assessment_type": r.assessment_type,
        "summary": r.summary,
        "percentiles": r.percentiles or {},
        "tags": profile.get("tags", []),
        "archetype": profile.get("archetype"),
    }


def compare(self_r: Result, other_r: Result) -> dict:
    """计算两人维度反差、契合度与模板结论。"""
    other = public_summary(other_r)
    self = public_summary(self_r)
    sp = self_r.percentiles or {}
    op = other_r.percentiles or {}
    keys = [k for k in sp if k in op] or list(op.keys())

    dims: list[dict] = []
    deltas: list[float] = []
    for k in keys:
        sv = float(sp.get(k, 0))
        ov = float(op.get(k, 0))
        d = round(sv - ov, 1)
        if abs(d) <= 5:
            v = "旗鼓相当"
        elif d > 0:
            v = f"你更{k}"
        else:
            v = f"对方更{k}"
        dims.append({"name": k, "self_pct": sv, "other_pct": ov, "delta": d, "verdict": v})
        deltas.append(abs(d))

    compat = 100 - round(sum(deltas) / len(deltas)) if deltas else 0
    compat = max(0, min(100, compat))
    verdict = _verdict(dims, compat)
    return {
        "self_summary": self,
        "other_summary": other,
        "dimensions": dims,
        "compatibility": compat,
        "verdict": verdict,
    }


def _verdict(dims: list[dict], compat: int) -> str:
    if not dims:
        return "数据不足,无法对比。"
    maxd = max(dims, key=lambda x: abs(x["delta"]))
    mind = min(dims, key=lambda x: abs(x["delta"]))
    return (
        f"你们在「{maxd['name']}」上反差最烈(差 {abs(maxd['delta'])} 分),"
        f"却于「{mind['name']}」最同频。契合度 {compat}%。"
    )

File: static/compare.js
// 关系对比 —— 粘贴对方对比码,渲染维度反差 + 契合度
// 复用 app.js:api(get/post 自动注入 X-User-Token)、getToken、injectLangSwitch

(function () {
  const codeEl = document.getElementById('code');
  const msgEl = document.getElementById('compare-msg');
  const outEl = document.getElementById('compare-result');

  function escapeHtml(s) {
    if (s == null) return '';
    return String(s)
      .replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;')
      .replace(/"/g, '&quot;').replace(/'/g, '&#39;');
  }
  function setMsg(t, kind) {
    msgEl.textContent = t || '';
    msgEl.className = 'compare-msg' + (kind ? ' ' + kind : '');
  }

  async function doCompare() {
    const code = (codeEl.value || '').trim();
    if (!code) { setMsg(mmI18n.t('compare.invalid'), 'err'); return; }
    setMsg(''); outEl.innerHTML = '';
    try {
      const data = await api.get('/api/compare?other=' + encodeURIComponent(code));
      renderCompare(data);
    } catch (e) {
      setMsg(mmI18n.t('compare.invalid'), 'err');
    }
  }

  function renderCompare(d) {
    if (!d.self_summary) { setMsg(mmI18n.t('compare.no_result'), 'err'); return; }
    const rows = (d.dimensions || []).map((dim) => {
      const sv = dim.self_pct, ov = dim.other_pct;
      const max = Math.max(sv, ov, 1);
      const sw = (sv / max) * 100, ow = (ov / max) * 100;
      return `
        <div class="cmp-row">
          <div class="cmp-head"><span>${escapeHtml(dim.name)}</span><span class="cmp-verdict">${escapeHtml(dim.verdict)}</span></div>
          <div class="cmp-bars">
            <div class="cmp-bar cmp-self" style="width:${sw}%"><span>${sv}</span></div>
            <div class="cmp-bar cmp-other" style="width:${ow}%"><span>${ov}</span></div>
          </div>
        </div>`;
    }).join('');
    outEl.innerHTML = `
      <div class="cmp-summary">
        <div class="cmp-card">
          <h4>${escapeHtml(mmI18n.t('compare.self_label'))}</h4>
          <p>${(d.self_summary.tags || []).map(escapeHtml).join(' · ')}</p>
        </div>
        <div class="cmp-score">
          <div class="cmp-ring">${d.compatibility}<small>%</small></div>
          <p>${escapeHtml(mmI18n.t('compare.compat'))}</p>
        </div>
        <div class="cmp-card">
          <h4>${escapeHtml(mmI18n.t('compare.other_label'))}</h4>
          <p>${(d.other_summary.tags || []).map(escapeHtml).join(' · ')}</p>
        </div>
      </div>
      <p class="cmp-verdict-text">${escapeHtml(d.verdict)}</p>
      <div class="cmp-dims">${rows}</div>`;
  }

  async function copyMine() {
    try {
      const list = await api.get('/api/me/results');
      if (!list || !list.length) { setMsg(mmI18n.t('compare.no_result'), 'err'); return; }
      await navigator.clipboard.writeText(list[0].id);
      setMsg(mmI18n.t('compare.copy_ok'), 'ok');
    } catch (e) {
      setMsg(mmI18n.t('compare.no_result'), 'err');
    }
  }

  document.getElementById('btn-compare').addEventListener('click', doCompare);
  document.getElementById('btn-copy').addEventListener('click', copyMine);
  codeEl.addEventListener('keydown', (e) => { if (e.key === 'Enter') doCompare(); });
})();
